src/plugins/data_source/server/saved_objects/credential_saved_objects_client_wrapper.ts

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {
+  HttpServiceStart,
+  SavedObjectsBulkCreateObject,
+  SavedObjectsBulkResponse,
+  SavedObjectsBulkUpdateObject,
+  SavedObjectsBulkUpdateOptions,
+  SavedObjectsBulkUpdateResponse,
+  SavedObjectsClientWrapperFactory,
+  SavedObjectsCreateOptions,
+  SavedObjectsUpdateOptions,
+  SavedObjectsUpdateResponse,
+} from 'opensearch-dashboards/server';
+
+import { SavedObjectsErrorHelpers } from '../../../../core/server';
+
+import { CryptographyClient } from '../cryptography';
+
+import { Credential } from '../../common';
+
+/**
+ * Describes the Credential Saved Objects Client Wrapper class,
+ * which contains the factory used to create Saved Objects Client Wrapper instances
+ */
+export class CredentialSavedObjectsClientWrapper {
+  public httpStart?: HttpServiceStart;
+
+  private _type: string = 'credential';
+  private _authType: Credential.SharedAuthType = 'shared';
+  private _noAuthType: Credential.NoAuthType = 'no_auth';
+  private _usernamePasswordType: Credential.UsernamePasswordType = 'username_password';
+
+  private _cryptographyClient: CryptographyClient;
+
+  constructor(cryptographyClient: CryptographyClient) {
+    this._cryptographyClient = cryptographyClient;
+  }
+
+  /**
+   * Describes the factory used to create instances of Saved Objects Client Wrappers
+   * for credential spcific operations such as encryption
+   * Check {@link Credential.CredentialSavedObjectAttributes} for attributes type details
+   * Check {@link Credential.CredentialMaterials} for credential materials type and
+   * credential materials content details
+   */
+  public wrapperFactory: SavedObjectsClientWrapperFactory = (wrapperOptions) => {
+    const createWithCredentialMaterialsEncryption = async <T = unknown>(
+      type: string,
+      attributes: T,
+      options?: SavedObjectsCreateOptions
+    ) => {
+      this.validateType(type);
+      this.validateAttributes(attributes);
+
+      const { title, authType, credentialMaterials, description } = attributes;
+
+      if (this._noAuthType === credentialMaterials.credentialMaterialsType) {
+        return await wrapperOptions.client.create(type, attributes, options);
+      }
```

Comment:
better to set username and password to undefined before saving

Reply:
> better to set username and password to undefined before saving

```
  private dropCredentialMaterialsContent<T = unknown>(credentialMaterials: T) {
    return {
      credentialMaterialsType: credentialMaterials.credentialMaterialsType,
      /**
       * The Saved object update API currently only accepts partial attributes as input.
       * Dropping the field or marking it as undefined will be ignored, and the existing field will remain unchanged.
       * TODO: Provide another update API to drop the field  https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2124
       */
      credentialMaterialsContent: {
        username: '',
        password: '',
      },
    };
  }
```